File: services/product-service-nest/src/modules/health/services/health.service.ts
import { Injectable, ServiceUnavailableException } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { InjectConnection } from '@nestjs/mongoose';
import { Connection } from 'mongoose';
import { ErrorCode } from '../../../common/constants/error-code.enum';
import { RedisService } from '../../../common/utils/redis.service';

@Injectable()
export class HealthService {
  constructor(
    @InjectConnection() private readonly connection: Connection,
    private readonly redisService: RedisService,
    private readonly configService: ConfigService
  ) {}

  async getHealth(): Promise<Record<string, unknown>> {
    return {
      status: 'ok',
      service: this.configService.get<string>('app.name', 'product-service'),
      timestamp: new Date().toISOString()
    };
  }

  async getReadiness(): Promise<Record<string, unknown>> {
    const mongoReady = this.connection.readyState === 1;
    if (!mongoReady) {
      throw new ServiceUnavailableException({
        code: ErrorCode.SERVICE_UNAVAILABLE,
        message: 'MongoDB is not ready'
      });
    }

    let redisReady = true;
    if (this.redisService.isEnabled()) {
      redisReady = await this.redisService.ping();
      if (!redisReady) {
        throw new ServiceUnavailableException({
          code: ErrorCode.SERVICE_UNAVAILABLE,
          message: 'Redis is not ready'
        });
      }
    }

    return {
      status: 'ready',
      dependencies: {
        mongo: mongoReady,
        redis: redisReady
      },
      timestamp: new Date().toISOString()
    };
  }

  async getLiveness(): Promise<Record<string, unknown>> {
    return {
      status: 'alive',
      service: this.configService.get<string>('app.name', 'product-service'),
      timestamp: new Date().toISOString()
    };
  }
}

File: services/product-service/internal/service/health_service.go
package service

import (
	"context"
	"net/http"
	"time"

	"product-service/internal/domain"
	"product-service/internal/httpx"
	"product-service/internal/timefmt"

	"go.mongodb.org/mongo-driver/mongo"
)

type HealthService struct {
	appName string
	mongo   *mongo.Client
	redis   *RedisService
}

func NewHealthService(appName string, mongoClient *mongo.Client, redisService *RedisService) *HealthService {
	return &HealthService{appName: appName, mongo: mongoClient, redis: redisService}
}

func (s *HealthService) Health() map[string]any {
	return map[string]any{"status": "ok", "service": s.appName, "timestamp": timefmt.ISO(time.Now())}
}

func (s *HealthService) Live() map[string]any {
	return map[string]any{"status": "alive", "service": s.appName, "timestamp": timefmt.ISO(time.Now())}
}

func (s *HealthService) Ready(ctx context.Context) (map[string]any, error) {
	mongoReady := s.mongo != nil && s.mongo.Ping(ctx, nil) == nil
	if !mongoReady {
		return nil, httpx.NewAppError(http.StatusServiceUnavailable, domain.ErrorCodeServiceUnavailable, "MongoDB is not ready", nil)
	}
	redisReady := true
	if s.redis != nil {
		redisReady = s.redis.Ping(ctx)
	}
	if !redisReady {
		return nil, httpx.NewAppError(http.StatusServiceUnavailable, domain.ErrorCodeServiceUnavailable, "Redis is not ready", nil)
	}
	return map[string]any{
		"status":       "ready",
		"dependencies": map[string]bool{"mongo": mongoReady, "redis": redisReady},
		"timestamp":    timefmt.ISO(time.Now()),
	}, nil
}
